### PR TITLE
Fix camera constraints with map padding

### LINKED ISSUE #8501: Camera padding is ignored by `defaultConstrain`, so the map cannot be panned to its edge
### Describe the bug

`Map#setPadding` documents the padding as the part of the canvas the camera treats as
visible, and `centerPoint` respects it. `defaultConstrain` does not — it measures against
the full canvas via `this.size`, so the pan limit is computed for an area larger than the
visible one.

With `padding.bottom = D`, the centre is drawn at `(H - D) / 2`, but the constraint still
reserves `H / 2` of map below the centre. The map's southern edge therefore stops `D / 2`
short of the padded viewport's bottom edge and cannot be brought into view.

This is the common "bottom sheet over a map" layout: the last `D / 2` of the map is
unreachable, and the amount lost changes as the sheet is resized.

### Steps to reproduce

1. Create a full-window map whose imagery spans the whole world.
2. `map.setPadding({ top: 0, right: 0, bottom: 500, left: 0 })` on a ~1000px-tall window.
3. Drag south as far as the map allows.

**Expected:** the map's southern edge can reach the bottom of the padded viewport (y = 500).
**Actual:** it stops at y = 750 — half the padding short. The strip between 500 and 750
can never be scrolled into the visible area.

Same applies to `left`/`right` padding horizontally, and to the `shouldZoomIn` branch,
which compares the bounds against the unpadded screen size.

### Cause

`src/geo/projection/mercator_transform.ts`, in `defaultConstrain`:

- `const {x: screenWidth, y: screenHeight} = this.size;` — unpadded canvas size
- `const shouldZoomIn = maxY - minY < screenHeight;`
- `const h2 = screenHeight / 2;` and the matching `minX`/`maxX` block

`this.size` is the raw canvas; `this._edgeInsets` is not consulted anywhere in the function.

### Suggested fix

Use the padded extent for these comparisons, so the constraint matches the region
`centerPoint` is derived from — i.e. `screenHeight - padding.top - padding.bottom` and `screenWidth - padding.left - padding.right`.

### Workaround

Supply a `transformConstrain` that reimplements the clamp against the padded size.

### Version

maplibre-gl 6.9.1; `defaultConstrain` is unchanged on `main` as of 2026-09-21.

### PR DESCRIPTION
**Pull request by erenbati**
## Summary

Respect map padding when constraining the camera to max bounds, and add a regression test for the padded viewport edge case.

Fixes #8501.

## Validation

Local tests could not be run in this environment because `github.com` could not be resolved during clone. The regression test is included; GitHub CI will provide the real validation.

## Launch Checklist

- [x] Confirm these changes do not include backports from Mapbox projects.
- [x] Briefly describe the changes in this PR.
- [x] Link to related issues.
- [x] Write tests for the bug fix.
- [x] Add an entry to `CHANGELOG.md` under `## main`.
- [x] Confirm I have read the MapLibre AI policy.

Assisted-By: ChatGPT (GPT-5.6 Sol)

### CODE REVIEW

File: src/geo/projection/mercator_transform.ts [hunk 1]
```diff
@@ -618,7 +618,9 @@ export class MercatorTransform implements ITransform {
         let maxX = worldSize;
         let scaleY = 0;
         let scaleX = 0;
-        const {x: screenWidth, y: screenHeight} = this.size;
+        const {top = 0, bottom = 0, left = 0, right = 0} = this.padding;
```
Comment (HarelM) [inline src/geo/projection/mercator_transform.ts:621]: How does this interact with handleEaseTo which seems to account for padding already?
Also, what about globe (vertical perspective) support for padding?

Padding is a hornets nest, so I find changes to it relatively risky...

**Comment by HarelM** [comment]
Thanks for taking the time to open this PR!

**Comment by clement-igonet** [comment]
Tried this with a sidebar that opens and closes. One case slipped through: `map.setPadding()` (and `jumpTo({padding})`) changes the padding but does not re-run the constraint, so the map stays where it was and the padded viewport can show the area outside the bounds until the next move.

Switzerland as `maxBounds`, red is outside the bounds. Top row: a 300 px sidebar is open with `padding.left = 300`, the map is panned west as far as the bounds allow, the border stops at the sidebar edge on both sides (this PR works). Bottom row: the sidebar closes with `setPadding({left: 0})`.



Left: the PR as is, the map does not move, so 150 px outside the bounds are now on screen. Right: with `constrainInternal()` added to `TransformHelper.setPadding`, the map slides 150 px and the border is back on the viewport edge. The same happens with zoom: when the padding shrinks and the bounds no longer fill the viewport, the zoom is not raised.

`easeTo` is fine already, since `interpolatePadding` calls `constrainInternal()`. The one-liner would be:

```ts
// src/geo/transform_helper.ts, setPadding
this._edgeInsets.interpolate(this._edgeInsets, padding, 1);
this.constrainInternal();
this._calcMatrices();
```

<details>
<summary>Unit test</summary>

```ts
test('re-constrains center when padding changes', () => {
    const transform = new MercatorTransform({minZoom: 0, maxZoom: 22, minPitch: 0, maxPitch: 60, renderWorldCopies: true});
    transform.resize(1000, 1000);
    transform.setZoom(7);
    transform.setMaxBounds(new LngLatBounds([-5, -5, 5, 5]));
    transform.setPadding({top: 0, right: 500, bottom: 500, left: 0});
    transform.setCenter(new LngLat(50, -50));

    transform.setPadding({top: 0, right: 0, bottom: 0, left: 0});

    const southeast = transform.locationToScreenPoint(new LngLat(5, -5));
    expect(southeast.x).toBeCloseTo(1000, 6);
    expect(southeast.y).toBeCloseTo(1000, 6);
});
```

Fails on the branch with `750`, passes with the change. geo, camera and map suites still green with it.
</details>

File: src/geo/projection/mercator_transform.ts (hunk 1 above)
Comment (erenbati) [reply src/geo/projection/mercator_transform.ts]: Thanks for the feedback. I checked both paths and updated the PR.

`handleEaseTo` is already covered: when padding is animated it calls `interpolatePadding()`, and `interpolatePadding()` already calls `constrainInternal()` on every frame. The missing path was the direct `setPadding()` / `jumpTo({ padding })` case, where the edge insets changed without re-running the constraint.

I added `constrainInternal()` to `TransformHelper.setPadding()` and added the regression case that clement-igonet found: after reaching a max-bounds edge with padding, removing the padding now re-constrains the center so the bounds stay aligned with the new viewport edge.

For globe / vertical perspective, the shared `setPadding()` change still goes through the projection-specific constrain implementation. `VerticalPerspectiveTransform.defaultConstrain` currently has the existing TODO to respect `_lngRange` / `_latRange`, so this PR does not try to introduce globe max-bounds support; it keeps padding changes consistent with the constraints each projection currently supports.

I also rebased the branch onto current `main` to resolve the conflict. The PR is now mergeable and 0 commits behind, and the full Tests workflow and CodeQL are green.
